### PR TITLE
Add optimisation for reflection-activated components to avoid adding a parameter in the resolve path if possible.

### LINKED ISSUE #118: Performance Degradation in version 9.0.0
**Describe the bug**
There is a significant performance degradation introduced in version 9.0.0 that was not present in 8.0.0. My traces led me to [KeyedServiceMiddleware line 22](https://github.com/autofac/Autofac.Extensions.DependencyInjection/blob/v9.0.0/src/Autofac.Extensions.DependencyInjection/KeyedServiceMiddleware.cs#L22) as suspect. As further evidence that the issue comes from this change, if I do not call `Populate` and just use `AutofacServiceProvider` directly, this line of code is not executed and the performance issues are not observed. This is obviously, not a suitable workaround, as it defeats half the purpose of this library, but is useful troubleshooting information.

These reproduction steps are in a benchmark, but we also witnessed this issue as about a 2x increase in CPU utilization and eventual application failure in a full fledged application while undergoing performance stress testing.

**To reproduce**
Reproduction of the issue, ideally in a unit test format.

* Wire up 1000 services in a dependency graph.
* Register each of these services with Autofac's `ContainerBuilder`. 
* Create a `ServiceCollection`, register some arbitrary service, and call `Populate` to load it into the container builder.
   * NOTE: The service loaded in this fashion isn't of importance, we just need the `KeyedServiceMiddleware` to be added to the resolve pipeline.
* Create an `AutofacServiceProvider` using the built container.
* Resolve the top-level service of this 1000 service graph in a benchmark.
* Run this benchmark for version 9.0.0 and prior versions (listed here are the old version we were upgrading from, as well as the version immediately below 9.0.0)

```c#
public interface ISomeInterface
{
    void DoWork();
}

public class OuterMostService(KayleighVaughan _dep) : ISomeInterface { public void DoWork() => _dep.DoWork(); }
public class OuterMostServiceShallow : ISomeInterface { public void DoWork() { } }

[Config(typeof(Config))]
public class AutofacBenchmarks
{
    private IServiceProvider _serviceProvider = null!;

    [GlobalSetup]
    public void Setup()
    {
        _serviceProvider = new AutofacServiceProvider(GetContainer().Build());
    }

    private ContainerBuilder GetContainer()
    {
        var services = new ServiceCollection();

        services.AddSingleton<OuterMostServiceShallow>();

        var builder = new ContainerBuilder();
        // not present in benchmark, but omitting this line results in 8.0.0 level performance
        builder.Populate(services);

        builder.RegisterType<OuterMostService>().AsSelf().AsImplementedInterfaces().InstancePerDependency();
        // Extension method, registers 1000 services in a tree structure
        builder.RegisterFakeServices();

        return builder;
    }

    [Benchmark]
    public void DirectResolveDeep()
    {
        _serviceProvider.GetRequiredService<OuterMostService>();
    }

    private class Config : ManualConfig
    {
        public Config()
        {
            var baseJob = Job.Default;
            AddJob(baseJob.WithNuGet([
                new NuGetReference("Autofac", "4.9.1"),
                new NuGetReference("Autofac.Extensions.DependencyInjection", "5.0.1"),
                new NuGetReference("Microsoft.Extensions.DependencyInjection", "3.1.14"),
            ]).AsBaseline());

            AddJob(baseJob.WithNuGet([
                new NuGetReference("Autofac", "8.0.0"),
                new NuGetReference("Autofac.Extensions.DependencyInjection", "8.0.0"),
                new NuGetReference("Microsoft.Extensions.DependencyInjection", "8.0.0"),
            ]));

            AddJob(baseJob.WithNuGet([
                new NuGetReference("Autofac", "8.0.0"),
                new NuGetReference("Autofac.Extensions.DependencyInjection", "9.0.0"),
                new NuGetReference("Microsoft.Extensions.DependencyInjection", "8.0.0"),
            ]));
        }
    }
}

public static class RegistrationExtensions
{
    public static void RegisterFakeServices(this ContainerBuilder builder)
    {
        builder.RegisterType<KayleighVaughan>().AsSelf().InstancePerDependency();
        builder.RegisterType<CastielQuintana>().AsSelf().InstancePerDependency();
        // 1000 services registered in this fashion
    }
}
public class KayleighVaughan(CastielQuintana _dep0, KeniaBennett _dep1, LeonardoFrost _dep2, PaulaGregory _dep3) { public void DoWork() { _dep0.DoWork(); _dep1.DoWork(); _dep2.DoWork(); _dep3.DoWork(); } }
public class CastielQuintana(TravisHuff _dep0, KarsynBradshaw _dep1, EmoryCrane _dep2, DellaFlynn _dep3) { public void DoWork() { _dep0.DoWork(); _dep1.DoWork(); _dep2.DoWork(); _dep3.DoWork(); } }
// ... more services wired up in a 1 parent, 4 children structure until I run out of dependencies, then the rest are leaf nodes.
public class SarahTruong { public void DoWork() { } }
public class AyanHobbs { public void DoWork() { } }
```

**Full exception with stack trace:**

```
No Exception - see benchmark results
```

**Assembly/dependency versions:**
```xml
<!-- See Config private class in C# code to multi-nuget-package version benchmarks -->
    <PackageReference Include="Autofac" Version="4.9.1" />
    <PackageReference Include="Autofac.Extensions.DependencyInjection" Version="5.0.1" />
    <PackageReference Include="Microsoft.Extensions.DependencyInjection" Version="3.1.14" />
    <PackageReference Include="BenchmarkDotNet" Version="0.13.12" />
```

**Additional context**
Benchmark results (sorry it's a screenshot, I don't have the console output on hand right now)
image

### PR DESCRIPTION
**Pull request by alistairjevans**
The issue this was raised against contains an _extremely_ deep resolve graph, but it does show there's a considerable overhead to changing the request parameters even if a type has no keyed service parameters.

Basically, when components are registering we'll "look ahead" at all the constructors for reflection activation, and if none of them have a `FromKeyedService` attribute, we can skip adding the additional parameter in the new middleware.

I've also moved the middleware to the start of the activation phase of the component pipeline, because there's no point adding the parameters if we've already instantiated a shared component.

## Point of Debate

@tillig, one question for you is in the review comment, about whether we need to support a certain esoteric use-case; if not then we get to stay fast in the delegate component path.

Closes #118 

### Benchmarks

This is the same slightly extreme benchmark used to raise #118 

v8
```
|            Method |     Mean |   Error |  StdDev |    Gen 0 |   Gen 1 | Allocated |
|------------------ |---------:|--------:|--------:|---------:|--------:|----------:|
| DirectResolveDeep | 677.2 us | 5.15 us | 4.56 us | 105.4688 | 34.1797 |    462 KB |
```

v9
```
| Method            | Mean     | Error     | StdDev    | Gen0     | Gen1    | Allocated |
|------------------ |---------:|----------:|----------:|---------:|--------:|----------:|
| DirectResolveDeep | 1.637 ms | 0.0378 ms | 0.1092 ms | 160.1563 | 56.6406 | 739.33 KB |
```

This branch
```
| Method            | Mean     | Error    | StdDev   | Gen0     | Gen1    | Allocated |
|------------------ |---------:|---------:|---------:|---------:|--------:|----------:|
| DirectResolveDeep | 670.4 us | 10.49 us | 10.77 us | 104.4922 | 27.3438 | 461.86 KB |
```

### CODE REVIEW

File: src/Autofac.Extensions.DependencyInjection/AutofacRegistration.cs [hunk 1]
```diff
@@ -82,12 +91,77 @@ public static class AutofacRegistration
...
+                {
+                    break;
+                }
+            }
+        }
+        else if (e.ComponentRegistration.Activator is DelegateActivator)
+        {
+            // With non-reflection activation, there are very few paths that would let the FromKeyedServicesAttribute
+            // actually work.
+            needFromKeyedServiceParameter = false;
```
Comment (alistairjevans) [inline src/Autofac.Extensions.DependencyInjection/AutofacRegistration.cs:139]: The only case I can find where someone might reasonably expect `FromKeyedServices` to be respected on a delegate activation is this one:

```
builder.Register(([FromKeyedServices("abc")] KeyedService service) => new ServiceConsumer(service));
```

Note that that's obviously not an MSDI registration, but it does kind of feel like something that _should_ work?

If we need to support that case, then we will have to add our middleware to all delegate activations, meaning that delegate activation slows down slightly.

If we don't need to support that case, then we can leave it false and never response `[FromKeyedServices]` on a delegate activation.

File: src/Autofac.Extensions.DependencyInjection/AutofacRegistration.cs [hunk 2]
```diff
@@ -63,11 +67,16 @@ public static class AutofacRegistration
...
     {
-        if (descriptors == null)
+        if (descriptors is null)
         {
             throw new ArgumentNullException(nameof(descriptors));
         }
 
+        if (builder is null)
+        {
+            throw new ArgumentNullException(nameof(builder));
```
Comment (tillig) [inline src/Autofac.Extensions.DependencyInjection/AutofacRegistration.cs:77]: Codecov wants a unit test to cover the null `builder` case.

File: src/Autofac.Extensions.DependencyInjection/AutofacRegistration.cs (hunk 1 above)
Comment (tillig) [reply src/Autofac.Extensions.DependencyInjection/AutofacRegistration.cs:139]: I'd say, for now, to skip this case and let delegates work fast. It's a pretty edge case thing, and it's absolutely doing the whole mix-and-match MEDI and Autofac registration types that we pretty clearly tell people _not_ to do. I didn't figure out a simple way to replicate anything like this in straight MEDI, so if mix-and-match doesn't work? I'm OK with that, at least for this.

**Review by tillig — CHANGES_REQUESTED** [review]
I think we can drop the delegate support to make this a touch faster; and if the unit test can be added to satisfy coverage, I think we're good to go.

File: src/Autofac.Extensions.DependencyInjection/AutofacRegistration.cs (hunk 2 above)
Comment (alistairjevans) [reply src/Autofac.Extensions.DependencyInjection/AutofacRegistration.cs:77]: Done. Added one for descriptors too.

**Review by tillig — APPROVED** [review]
🐲

**Pull request merged**